### ISSUE TITLE
[feature request] Implement torch.stack() / _cat() for also for torch.HalfTensor

### ISSUE THREAD

**Issue by timothygebhard** [pytorch/pytorch#6968]
## Issue Description
It seems that stacking is currently (PyTorch 0.4) is not supported for tensors of type `HalfTensor`; see code example. 

I don't know if this is intentional for some reason (hence I didn't label this issue as a bug), but I couldn't find any issue or documentation pointing to that. Since some methods, like `torch.utils.data.DataLoader`, depend on `torch.stack()` (in this case implicitly through the `default_collate()` method), it would be great if either stacking for HalfTensors could get implemented, or the documentation could be updated to contain some kind of hint.

## Code Example
```
Python 3.6.5 (default, Mar 30 2018, 06:41:49)
Type 'copyright', 'credits' or 'license' for more information
IPython 6.3.1 -- An enhanced Interactive Python. Type '?' for help.

In [1]: import torch

In [2]: torch.__version__
Out[2]: '0.4.0'

In [3]: tensor_1_float = torch.tensor([1, 2, 3], dtype=torch.float)

In [4]: tensor_2_float = torch.tensor([4, 5, 6], dtype=torch.float)

In [5]: torch.stack([tensor_1_float, tensor_2_float])
Out[5]:
tensor([[ 1.,  2.,  3.],
        [ 4.,  5.,  6.]])

In [6]: tensor_1_half = torch.tensor([1, 2, 3], dtype=torch.half)

In [7]: tensor_2_half = torch.tensor([4, 5, 6], dtype=torch.half)

In [8]: torch.stack([tensor_1_half, tensor_2_half])
---------------------------------------------------------------------------
RuntimeError                              Traceback (most recent call last)
<ipython-input-8-10f49aa85ea5> in <module>()
----> 1 torch.stack([tensor_1_half, tensor_2_half])

RuntimeError: _cat is not implemented for type torch.HalfTensor
```

On PyTorch 0.3.1, one gets basically the same error when one tries to do the same thing:
```
TypeError: Type torch.HalfTensor doesn't implement stateless method cat
```

**Comment by soumith** [comment]
we dont support any operation other than `copy` and serialization on torch.HalfTensor (the CPU version). The GPU HalfTensor supports all operations

**Comment by apaszke** [comment]
@soumith while we don’t want to have any math operations for half on CPU I feel like the argument given by @timothygebhard for having cat is very reasonable. Cat can be implemented efficiently because it doesn’t care how the data looks like, it’s only copying raw bytes around.

**Comment by ezyang** [comment]
If you do this, it would probably be a good idea to port THTensor_(cat) to ATen while you're at it.

**Comment by towr** [comment]
I ran into this issue with a `DataLoader` that works on a `numpy.fp16` dataset. 
Writing a custom `collate_fn` for the dataloader solves the immediate problem, but it would be nice if it works out of the box.

**Comment by TimZaman** [comment]
I'm confused, why did we close this issue? There was an MR for this that looked like it worked, but was subsequently closed, together with this issue/

**Comment by soumith** [comment]
as replied on the PR, it's merged into master, it's available in the nightlies, and will be part of 1.1
